### ISSUE TITLE
Web support via Puppeteer

### ISSUE THREAD

**Issue by awinograd** [wix/Detox#1882]
**Is your feature request related to a problem? Please describe.**

We deploy a react native web app in addition to ios/android. Ideally we'd be able to test the app using the same harness as our native RN apps.

**Describe the solution you'd like**

I've started implementing a POC `PuppeteerDriver` and web specific actions/matchers. I have some basic assertions / waitFor / actions working (though the code is currently a mess). If this is something your project is interested in having in master, I'd like to discuss approach / implementation details to come up with something that'd be acceptable.

**Describe alternatives you've considered**

There are alternative frameworks (e.g. Cypress) that are specific to web. Test could be 2x written for the web platform specifically.

I could stub the detox globals with ones backed by cypress / puppeteer without integrating directly into detox. Maintain the api, but have a different implementation.

**Comment by LeoNatan** [comment]
Hello,

Since we have no way of supporting or maintaining this in the future, I think this would be best handled as an external driver, which you can maintain properly.

Do you know what is needed to support such a driver in Detox currently? We will be willing to accept a PR with API to be able to provide your own driver, which will then open the door to third party drivers. 

What do you think?
Thanks

**Comment by awinograd** [comment]
@LeoNatan, I understand not wanting to support the plugin long term. Supporting external drivers sounds like a great path forward to me. Do you think that `configuration.type` could be set to a path on disk to a driver and change 

https://github.com/wix/Detox/blob/444b8d30fe32afab4b2811b84ef4297bb18983b6/detox/src/Detox.js#L55

to something like:

```
    const DeviceDriverClass = DEVICE_CLASSES[this._deviceConfig.type] || require(this._deviceConfig.type);
```

**Comment by LeoNatan** [comment]
@noomorph @rotemmiz Can you pls advise? Thanks

**Comment by rotemmiz** [comment]
We already had some discussion about such an API here https://github.com/wix/Detox/issues/1091, though I don't remember the details (I will read the thread again later tomorrow).
I would really be glad if this is pushed forward!

**Comment by awinograd** [comment]
Thanks for linking the relevant issue! The proposed api in there looks like it would work. I've started an initial investigation into trying to add support and registering the driver + extending the base class are definitely the two main things i've run into.

My implementation is moving forward... Got a lot of the basic stuff working (most matchers, most assertions, screenshot plugin, video recording plugin in progress... etc). Code isn't in the best condition, but my first approach was to get something that works and then hopefully cleanup from there.

If you're interested in what I have so far, here's a diff:

https://github.com/wix/Detox/compare/master...ouihealth:aw-puppeteer?expand=1

**Comment by awinograd** [comment]
Running list of other issues / incompatibilities i'm seeing with a custom driver:

1. Notion of binaryPath doesn't apply. Closest proxy is IMO the domain. This matters when setting permissions since they're scoped to domain.
1.

**Comment by awinograd** [comment]
@rotemmiz ,

I have most of the e2e tests running on web with my detox fork.  The remaining i've skipped because they don't apply to web or are too complex to implement at the moment. I also have some questions around certain functionality to make sure I support it in a similar way. For example, I dont really understand the timeout tests. It seems like some timeouts are ignored while others aren't. 

I haven't yet done any of the work to extract into a separate repo / module. Do you think the API proposed in #1091 is a good place for me to start? If so, I can start the work to get to something similar to that.

Thanks for all your guidance

**Comment by rotemmiz** [comment]
Hey @awinograd, I haven't forgotten you, I was just unable to dive into the implementation until now. I will have a better answer by tomorrow.

**Comment by rotemmiz** [comment]
So this is an initial draft I came up with, not sure if its full though, I'll revisit it if needed.

### package.json

```js
"ios.sim.release": {
        "binaryPath": "...",
        "build": "...",
        "type": "my-custom-driver-package-name",
        "device": {
	  // can be pretty flexible (according to the device driver), add whatever device specification needed
        }
      },
```

### Detox.js

Add an ability to load a dynamic driver from `my-custom-driver-package-name`  https://github.com/wix/Detox/blob/master/detox/src/Detox.js#L55

### API and Globals 

By default Detox loads `device`, `expect`, `by`, `element` and `waitFor` as globals, but it can be skipped and given as an option to the user to load explicitly.
https://github.com/wix/Detox/blob/master/detox/src/DetoxExportWrapper.js#L23


### Reusing Detox internals 

Not really sure what is reusable, simple things like `DetoxRuntimeError`, or `environment.js`, maybe `DeviceRegistry` for parallel test execution support, maybe `argparse` as well.
Exporting all those might not be a huge effort, but I wouldn't want them to be moved out of the codebase, but rather exported so they'll be accessible like so

```js
const argpargs = require('detox').argparse;
```

or something similar.

This will require `detox` to be a devDependency of your plugin. I think that this is an OK requirement, but we can discuss this.

Regarding artifact collection, is that something you are planning on doing? 

Did I miss anything @d4vidi @noomorph ?

**Comment by awinograd** [comment]
Thanks for the draft @rotemmiz.

Great to know I can stuff things into the `device` key of the config. However, do you have a good idea for what to do with `binaryPath`. Should I just leave it blank? I'm currently using it as the base url to load in the browser, but it expects an absolute path which means i'm currently prefixing a `/` to my url. e.g. `/https://example.com`

> Add an ability to load a dynamic driver 

In the linked PR I saw an API to add the driver to the DEVICES hash. That seems like a good path forward to me.

> API and Globals

From what I have so far,  I think this can be handled in the same way as current drivers.

> Reusing Detox internals

Haven't looked into using any of the utilities you mentioned. But I have used things like `Client` directly. Right now I dont support parallel test execution but definitely something I'd like down the line. 

It would be nice to bundle the `test/e2e` files with the detox package (or publish as a separate package) so I can run my driver against the same set of tests as normal detox.

I currently have support for screenshots in my driver, as well as video recording when running in non-headless mode. Haven't yet been able to get headless videos to work (even using something like xvfb).

**Comment by awinograd** [comment]
Just spent some time trying to move my driver into it's own package...

https://github.com/wix/Detox/compare/master...ouihealth:aw-add-driver?expand=1
https://github.com/ouihealth/detox-puppeteer

The above detox fork and `detox-puppeteer` repo work for me right now! detox-puppeteer is using a lot of internal modules, but doesn't seem to have an issue requiring them from nested paths

In my `e2e/init` I added:

```js
import { addDriver, cleanup, init } from 'detox'; 
addDriver('web.puppeteer', PuppeteerDriver);
```

Still likely have some cleanup to do before publishing to npm but it's looking promising!

**Comment by awinograd** [comment]
More progress... publish some NPM packages that work together.

1. A fork of detox (linked in the previous comment) with `addDriver` exposed https://www.npmjs.com/package/oui-detox
1. https://www.npmjs.com/package/detox-puppeteer

`oui-detox` is just a temporary package to show that the integration is working. Ideally we could get the PR merged once we're happy with the API / exports / etc and delete the `oui-detox` package

**Comment by awinograd** [comment]
Note sure if it makes sense to ask this here, or open a new issue... but i'm running to issues with Artifacts due to `DetoxAdapterImpl`.

I'm seeing `beforeEach` for the next test get called before the current test's `afterEach` due to `testComplete` enqueuing the afterEach call to be run later. Can't find anything in the git history that explains what the enqueuing/ flushing is for

**Comment by awinograd** [comment]
Hello. Sorry for the semi stream of consciousness last week with. The driver has been stable me on CI the last week. Let me know if you'd like to me open https://github.com/wix/Detox/compare/master...ouihealth:aw-add-driver as a PR. I was planning on it, but it said to get approval in an issue, first

**Comment by rotemmiz** [comment]
Re the test lifecycle, what test runner are you using? mocha/jest-jasmine/jest-circus?
@noomorph/@d4vidi  can you share some of your experience re test runner lifecycle?

**Comment by rotemmiz** [comment]
@awinograd re the PR, start by opening in and let's discuss in the PR itself (reference this issue in it).
I already have some questions regarding that API, `addDriver(name, driver)` seems awkward to me since it needs to run in the test context (rather than the cli/configuration context).

What I suggested in https://github.com/wix/Detox/issues/1882#issuecomment-584090146 looks like a better solution to me since it separates the test configuration, and will only load your custom driver on your custom configuration instead of doing it on all of them.

**Comment by awinograd** [comment]
> Re the test lifecycle, what test runner are you using? mocha/jest-jasmine/jest-circus?

I'm using jest-jasmine, but the "issue" appears to be in DetoxAdapterImpl which is shared by jest jasmine and circus

> I already have some questions regarding that API, addDriver(name, driver) seems awkward to me since it needs to run in the test context (rather than the cli/configuration context).

I originally had trouble getting your suggested API to work  but turns out that was because I was using `yarn link` which messes up resolving of other node_modules since the file source isn't actually in the project node_modules. Just got it to work via published packages, though. I'll open the PR.

**Comment by awinograd** [comment]
Support for this released in v16.3+!
